### ISSUE TITLE
Please release java package to maven central

### ISSUE THREAD

**Issue by bond-** [ldthomas/apg-java#6]
Looks like maven central doesn't hold this artifact, please release it to maven central.

**Comment by ssill2** [comment]
I'd be very intersted in this as well.  I have a local copy of the source that I've made into a maven project for local use, but it would be great if this project were converted to maven.  This opens up a lot of possibilities, such as posting to maven central repo.

**Comment by ldthomas** [comment]
Here's the problem. I haven't written a line of Java code since I published this 7 years ago. Until this issue I had never even heard of maven. So I probably won't get around to learning how to using it anytime soon.  But thanks for your interest and pointing this out to me.

**Comment by ssill2** [comment]
The process to make this a maven project is relatively straight-forward.
No code has to be changed at all.  I could create a fork and do the change
in my fork if you're interested in seeing what it looks like.

**Comment by ldthomas** [comment]
OK, let me see what it looks like. I've got a little break coming up and I'll see if I can get up with maven.

**Comment by bond-** [comment]
@ldthomas Let me know if need a hand, I can write something up. But please do fix #5 before releasing it to Maven.

**Comment by ssill2** [comment]
> OK, let me see what it looks like. I've got a little break coming up and I'll see if I can get up with maven.

Yep, I've already got this work done locally since that's how I use your stuff lol.  All the rest of my java projects are maven so that was the easiest way to do it.

Give me a day or so since I'm right in the middle of something and then I'll create a fork and make the changes for you to inspect.  I agree with bond though.  it would be good to update the license prior to posting it to maven central repo.

**Comment by ssill2** [comment]
> OK, let me see what it looks like. I've got a little break coming up and I'll see if I can get up with maven.

https://github.com/ssill2/apg-java

Ok,  here's the project converted to build with maven.  I left your build dir in place, but it's not needed anymore.
Simply doing "mvn clean install" will do the following:
Build the code
Build javadoc

Maven by default writesthe compiled jar under target/ and the javadoc under target/site


Maven mostly works by convention so any junit tests would live under src/test/java
Any resources needed by unit tests would live under src/test/resources
For the actual code, maven uses src/main/java and src/main/resources
In the pom.xml you can define other folders to pull code and resources from, but the maven-compiler-plugin plugin just knows the above structure out of the box.

Many maven plugins run automatically so running the "mvn clean install" would actually build the code and run any unit tests present.

As I said, none of your source had to change I just moved things around so they would be where maven expects to find them.
Maven is really awesome especially if you project needs other third party deps.  Versioning is just part of a maven project.

Anyway, let me know what you think.

**Comment by ssill2** [comment]
I should add that the way I ended up using your apg tool was to write a maven plugin that used the version of your project that I mavenized as a dependency and allowed me to pass in the path to the .bnf files etc.   This allowed me to generate java parser from my .bnf files.

**Comment by bond-** [comment]
@ssill2 I ended up using antlr because I wasn't able to figure out apg-java, would be helpful if you can share an example app on how you used the syntax tree.

**Comment by ssill2** [comment]
I set this up quite some time ago.  We get a file with expressions in a specific syntax and this lib is what we ended up using.   As I had mentioned, in order to call this lib from another maven project where our .bnf files exist and have java code generated we wrote a maven plugin that calls this lib with the arguments needed, such as java package of the resulting generated classes.  Doing it this way allowed us to add the apg-java project as a dep and invoke its classes during the build of another project that produced the generated java code.

**Comment by ssill2** [comment]
Oh you mean to use the generated grammar?

**Comment by bond-** [comment]
@ssill2 I meant the last part of your process 🙂 , how did you use the generated java code?

The generated java code should produce an AST, how did you walk that tree to accomplish what you wanted to do?

![image](https://user-images.githubusercontent.com/776647/116239803-14416c00-a731-11eb-9a4f-89294ad519df.png)
Source: https://github.com/ldthomas/apg-java/blob/32000d88be16f297d04e8908f6e27daa985ed0a6/README.md

**Comment by ssill2** [comment]
Before I knew about this tool I had written another domain specific language we use in our system using javacc.    So you must do is write a parser that uses the grammar generated from the .bnf file.  The parser merely intercepts the grammar keywords as it gets them and allows you to do something with them.
so if you have bnf written for a language like:
this equals "that"

your bnf would say that "this" is a keyword, "equals" is a keyword and  quotes with text in the middle is a value

your parser would  simply have handlers for when those things were seen.  In my case I have a StringBuilder that I'm appending to as I build the expression from the keywords I see.


In my case do this.
Generate the grammar.
The parser uses the grammar to generate expressions in the language I wrote using javacc.   This intermediate language is then translated to java's EL to be actionable.

**Comment by ssill2** [comment]
I had to go back and look at the code again, it's been a couple years since this was implemented.

You create a new instance of apg's Parser object
Parser p = new Parser(YourGrammar.getInstance());;

You set the starting rule
                p.setStartRule(YourGrammer.RuleNames.YOUREXPRESSION.ruleID());
                ast = p.enableAst(true);
                for (int i = 0; i < YourGrammar.ruleCount; i++)
                {
                    ast.enableRuleNode(i, true);
                }

You add your rule callbacks, as many as are needed.

Ast.AstCallback acb = new Ast.AstCallback(ast)
        {
            @Override
            public void postBranch(int offset, int length)
            {
                // do something when this rule is encounter
               // like add some text to a stringbuilder
            }
        };
ast.setRuleCallback(YourGrammer.RuleNames.YOUREXPRESSION.ruleID(), acb);


Set input string in your parser
p.setInputString("this equals that");

Run the parse
Parse.result result = p.parse();

Check for success
boolean success = result.success();
if(success)
{
// call getter on your parser implementation to get the result, whatever that is.  in my case, it's another expression in a different language.
}

**Comment by ssill2** [comment]
if you look at the inifile in the examples package this basically demonstrates the concept.

**Comment by ssill2** [comment]
The hardest part is coming up with your bnf file and then tying the rules together.   Sometimes your parser has to keep state between rules.
for instance if you have an expression
header.subject contains "sometext"

you have to keep track of the field(header.subject), the operation(contains), and the value being looked for.  So your parser may have a few instance vars like Stacks and Lists to hold the info between rule evaluations.  And these have to be cleared in some cases.

In my case I have one project that invokes apg to generate the grammar, and another that defines the parser and unit tests.  This way the second project can have a regression tests that invoke the parser with various expressions and validate that the correct result is produced by the parse.

**Comment by ssill2** [comment]
My question for ldthomas is.  From looking at your page about apg-7 it seems like even though the tool is written in C, it can still produce java grammar file, is that correct?   If that's the case it should be possible to build that once and then invoke it from a maven build via maven-exec-plugin to produce the same sort of thing I'm currently doing with this project.

**Comment by ldthomas** [comment]
You guys are getting way ahead of me, but that’s cool. Like I said, I’m really, really rusty with Java and know zip about maven. Looks like maven is basically npm for Java. Or something close to that.

@ssill2 I see how you rearranged the project for maven. That looks easy enough but the hard part is how to install maven and make it work. I’m pretty sure I can figure that out but hints are accepted and it won’t happen overnight.

Sorry, but APG-7.0 is C only. But changing the license to BSD is a go. I'm closing #5.

@bond For generating and walking the AST see [examples/demo](https://github.com/ldthomas/apg-java/blob/master/src/examples/demo/TranslateAst.java). It has a pretty straightforward example of doing it. I can’t remember why there is a separate call to enable a node and to set a callback for a node. I think later versions in other languages combine these two.

To see how the Ast class walks the tree see the translateAst() member in [Ast](https://github.com/ldthomas/apg-java/blob/master/src/apg/Ast.java). It’s just a depth-first walk through the nodes you choose to keep, calling your translation (callback) functions at each node, down and up.

**Comment by ssill2** [comment]
Yeah, maven is just a central place to pull deps from.

Maven is trivial to install.  pull the .zip from maven.apache.org unzip it somewhere and add that path/bin to you PATH and that path as M2_HOME

if you have an IDE that integrates well with maven like netbeans does,  it can easily search for and add deps to your projects.   I think the others can be made to, but I was never happy with how eclipse integrated with maven.  


> 
> Sorry, but APG-7.0 is C only. But changing the license to BSD is a go. I'm closing #5.
> 
Ah too bad about 7 being C only.  I guess 1.x line is where it's at for me then.  If you do end up bringing in my changes can we go with a 3 component version, i.e. 1.0.0 ?   This gives you some flexibility in tags for major.minor.patchlevel fixes.

**Comment by ssill2** [comment]
Also, if you need some examples of adding junit tests let me know.  Maven makes it super easy to run your tests at build time and fail the build if they don't pass.  you don't explicitly need to add the maven-surefire-plugin unless you want to configure something specific or different from the defaults.

**Comment by ldthomas** [comment]
OK, I have maven installed and working and I'm trying to integrate it with an IDE. But I'm having the same problem with both Intellij and NetBeans. In my Hello World project I see the error "cannot access java.lang". I've been scouring the Internet and always get that the fix is in Project->Properties->Libraries. But with my recently installed netbeans 12.3 there is no Libraries category. Only "JavaScript Libraries" which is blank. Makes no sense. @ssill2 Have you had this problem or know the solution?

**Comment by ssill2** [comment]
Alright, I was going to mention this but I forgot.

Which java version are you using?  8 or 11?

I just recently switched all of my work projects to 11, which is the latest LTS release of java.  

I'm using java 11

So here's how I set up my environment to build java code
ssill2@centos8-dev ~]$ cat .bash_profile 
if [ -f ~/.bashrc ]; then
	. ~/.bashrc
fi

M2_HOME=/home/ssill2/dev/apache-maven-3.6.2
JAVA_HOME=/usr/lib/jvm/java-11-openjdk
EDITOR=vim
PATH=${JAVA_HOME}/bin:${M2_HOME}/bin:${PATH}
export M2_HOME JAVA_HOME PATH



You can then verify which java you're running
[ssill2@centos8-dev ~]$ which java
/usr/lib/jvm/java-11-openjdk/bin/java

I can then cd to the git checked out dir
and run "mvn clean install" to get

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  8.224 s
[INFO] Finished at: 2021-04-28T15:04:26-06:00
[INFO] ------------------------------------------------------------------------
[ssill2@centos8-dev apg-java]$

**Comment by ssill2** [comment]
The first time you run the build it will download a bunch of stuff and cache it under ~/.m2/repository

Future builds will be faster since the dependencies will be cached.

**Comment by ldthomas** [comment]
That all works, but when I open the project up in NetBeans I still get "cannot access java.lang" and I can't run the project from the IDE.

**Comment by ssill2** [comment]
Standby let me pull up my fork in netbeans.  What netbeans version are you using?

**Comment by ssill2** [comment]
I was able to open in Netbeans 12 using java11

![image](https://user-images.githubusercontent.com/25040169/116475781-8a2ffb00-a837-11eb-8231-6e0222b9ad50.png)

**Comment by ssill2** [comment]
If you are using a version of netbeans before 10.x you'll need to upgrade.  8.x was the last version released by oracle and did NOT support java 9+.

**Comment by ssill2** [comment]
for cases where I need to support java 8 I usually have a branch of the project with the maven-compiler plugin configured to build as java 8.  There's a multitude of ways to do this, you can also do maven profiles to accomplish this.  I just went through the effort to get all my work projects off of java 8 since only RedHat is still maintaining java 8 and most of the other tools I use have moved on to java11.

**Comment by ssill2** [comment]
also, when you say "run the project" do you mean build?   Or are you expecting some particular class with a main in it to run?

**Comment by ssill2** [comment]
If you want I can add something to pom to allow you to run the under examples.  Those are things that in a maven project I would probably implement as junit tests, but I can give you a way to run that main in examples.Main

**Comment by ssill2** [comment]
just added a unit test that invokes your demos when you do "mvn clean install"

To run the build without running unit tests do "mvn -DskipTests clean install"

**Comment by ldthomas** [comment]
I've gotta run. Been a long day. I'll get back to you tomorrow. I'm using /usr/lib/jvm/java-11-openjdk-amd64/bin/java with NetBeans 12.3. I'm trying to execute (not build) a simple Hello World app right now. I looks like NetBeans is using JDK 11(Default) which has no libraries. I'm trying to configure it to use just JDK 11, which seems to have all the libraries. So far no luck but I'll give it a fresh look tomorrow.

Actually, once I'm up and running with this I was thinking that the examples should be moved into the tests, probably with a few modifications to generate whatever it is that tests are expected to generate.

**Comment by ssill2** [comment]
No worries, I can assist you tomorrow.   Netbeans is hands down the best ide for working with maven.  it just natively understands the pom.xml files and doesn't need to create a bunch of ancillary files from the pom.

I was going to say that the best thing would be make individual unit tests out of those demos but I wanted to just verify I could run your demos at build time.   Cheers

**Comment by ldthomas** [comment]
Well, after an entire morning of exasperation, seeing tutorials where everything seems to work out of the box and lots of suggestions to modify /etc/netbeans.conf, which had no effect, I finally found the answer I was looking for on stackoverflow. Tools->Options->Java->Maven let's you chose the maven and JDK versions of your choice. My default installation had a default JDK that pointed to /usr. Huh? Anyway, I'm now able to run, debug, unit test and build the jar for my simple Hello World project. I need to do some deep diving into the maven pom and JUnit. Maybe I can actually get Java APG maven configured and working sometime soon. I've looked at your fork and it's starting to make sense.

I'm busy this PM but will be back at it soon.

**Comment by ssill2** [comment]
That's quite interesting.  it's why I always make sure JAVA_HOME is set to something I know.  With the switch to java 11, it was imperative to make sure I was pointing at the correct java. Your stuff doesn't make any use of things like jaxb, but a lot of my stuff does.  jaxb was removed from the default jdk in java9+ and it's expected to be supplied via normal dependencies from then on.  Glad you got it working!  if you look at my branch there is a junit test to use as an example.  Maven just runs these automatically on build unless you supply -DskipTests

Let me know if you run into anything else!

**Comment by ldthomas** [comment]
@ssill2 Here’s a question for you. You’re high on NetBeans so maybe you know the answer to this. I’ve been looking for hours and I’m stumped. I can create a NetBeans maven project and run and debug it. I’ve found that by clicking the double yellow arrow in the Output window I can pass arguments to  main() on the exec.args= line. What I can’t figure out is how to pass arguments to main() when running the debugger. Seems like such a primary essential task but I can’t find a way or an answer anywhere.

I have apg.jar installed in the local repository and I’m trying to give instructions on how to use it to run, modify and debug the examples in a NetBeans project. I’ve got it all figured out except for passing arguments to main() with the debugger.

Next up I have to figure out how to deploy apg.jar to Maven Central. That looks like its going to be lots of fun too.

**Comment by ssill2** [comment]
The way I usually do this is set up a unit test that runs the main with the args. 

this way you can put your cursor in the particular test method and then right click and say debug focused test method.  it's the easiest way and works great.  you might need to change the pom.xml to make sure the maven-compiler-plugin has debug set to true.

**Comment by ssill2** [comment]
if you look at my fork where I invoke the demos in from a junit test you can do it like that.   if you right click in a package under unit test source code you can do a new Junit test and it will rough one out for you.

**Comment by ldthomas** [comment]
OK, got it. That works. Seems like a round about way of doing it, but I can live with it. Still, I'm sort of dumbfounded that a mainstream IDE doesn't have a way to pass arguments to main() in the debugger.

**Comment by ssill2** [comment]
technically you can.  if you right click on your project in netbeans and open properties, you can put parms in the run section.   this would be the main class, etc.  once you've set those you can right click on the project and debug it.  I like the unit test way of doing it though.  The thing I like about netbeans it's that it's very minimalistic as far as creating project configuration ancillary files outside the pom.xml
Eclipse and itellij like to litter your workspace with .classpath dirs and .project files etc.

**Comment by ssill2** [comment]
you can definitely use eclipse or intellij if that's your preference but I'm partial to netbeans because of the close maven integration.

**Comment by ldthomas** [comment]
Darned if that doesn't work - the right click on Properties. That's the solution I just spent about 3 hours looking for.

**Comment by bond** [comment]
Hmm, always get tagged in all sorts of random issues :P bond with a dash behind it? Really? :P

**Comment by ssill2** [comment]
> Hmm, always get tagged in all sorts of random issues :P bond with a dash behind it? Really? :P

could be worse I suppose!

**Comment by bond-** [comment]
> Hmm, always get tagged in all sorts of random issues :P bond with a dash behind it? Really? :P

Had to be creative to find good usernames

**Comment by ldthomas** [comment]
I’ve spent the morning looking at how to deploy to Maven Central and there are a lot of hoops to jump through. A lot more work than I really want to get involved with right now. I’ve got other things I need to get on to.

@ssill2 I’ve cloned your fork and it is a nice set up. I definitely see the logic of it and now that I’ve got pretty good control of NetBeans, it is a nice development environment. But there are a couple of minor problems with it. 

I added a test package to run apg.Generator.main() as well and that test fails with the /help option. The build error indicates that the SureFire plugin failed. Subtle problem: the generator uses System.exit(0) to exit a constructor on /help or a bad argument and SureFire is incompatible with System.exit(). (https://github.com/bmwcarit/barefoot/issues/47). System.exit() is a valid and probably the best way to exit from deep in the call stack, so it doesn’t make sense to program a workaround just to satisfy a SureFire bug. 

The other problem is that your project builds a jar file with apg _and_ the examples in it. I’d like to keep those separate, especially if apg.jar is to be deployed eventually. You probably know how to fix that, but it is time for me to move on. 

Also, I don’t really want to make a major change to the directory structure of the current repository and force all of those forks out there to conform to a maven layout. Looks like maven is the way to go if you are a Java developer, but it is also pretty clear that there are other options and probably not everyone wants it laid out that way.

But here is what I propose. I’ve made all of the changes to convert it to 2-clause BSD along with a few other tweeks. I’ll add instructions on how to install apg.jar in the local maven repository and how to include it in a project that needs it. That means if you want to include apg.jar in your maven project, you just have the extra step of cloning apg-java and doing a maven install. But after that using it in your project is the same as if it came from Maven Central. At least that’s the way it looks to me right now.

I’ll try to get that done and push it up to a branch (branch name maven, naturally) today. You can take a look at it and we can go from there. I’ll be occupied all day Friday and Monday but I really need to wrap this up soon.

**Comment by ssill2** [comment]
Fair enough.

You can absolutely run multiple executions of the maven-jar-plugin and have it include the examples package in a different jar.  Maven is extremely flexible like that.  it would be interesting to poll folks using this to see if they'd object to a maven conversion in general.  Your options are mainly ant and gradle.  I've not yet jumped on the gradle bandwagon and I used to be a big fan of ant before maven.  Clearly if start out with a maven project from the beginning you sort of structure things in a maven-friendly way.  Things like your examples would likely be created as test code under src/test/java and thus wouldn't end up in the final .jar anyway.

I appreciate you spending the time on this.

 So your maven branch won't actually be a maven project if I'm understanding correctly.  In that case would you still merge changes from master there or would that branch effectively be dead at that point?  If you're making this BSD license would you have an issue with me making a fork and maintaining the fork?

So up until talking to you I had assumed this project was abandoned, and since it wasn't on maven and I did build code to use it, I was maintaining a local copy of it.  My job requires me to use Fortify SCA to scan the code and remediate anything it finds.  Fortify scans on APG return a bunch of mostly nonsense findings, but I still have to answer for them.  Up until recently a response of "this is a third party lib and I don't have control over the code" was sufficient.   They'd really like me to see about addressing some of them.   I really the job you did on this thing and I'd love to keep using it.

**Comment by ldthomas** [comment]
maven was probably a bad choice for the branch name. I only wanted to push it up as a branch for review and discussion before merging it into master. Except for the license change and a few tweeks to the README file it is the same as 1.0. However, all of the files have changed as I have reformatted them with NetBeans' nifty formatting tool (I think there were a lot of irritating tabs in the originals), added a copyright and license notice header and regenerated all of the grammar.java files for all of the .bnf files, including ABNFforSABNF.bnf, since the generator now adds a header and specifies version 1.1.0.

I have no problem with you maintaining your own fork. I think that anybody can create a fork and do with it what they like. No problem there.

It's always nice to hear that someone out there is actually using your code. If you find any security problems with APG I would love to hear about them. No promises on fixing anything, but still.

I hope to have the branch pushed today but I'll be signing off after 5:00PM EDT.

I'm sure there will be more to discuss, but I just wanted to thank you for all the help you have given me. It's been great fun.

**Comment by ssill2** [comment]
sounds good!   I'll keep an eye out for it.

**Comment by ldthomas** [comment]
maven branch is up.

**Comment by ssill2** [comment]
alright, thanks.  I'll give this a look.

**Comment by ssill2** [comment]
So I reforked from your maven branch and converted it to maven again.

I ran a scan against it with Fortify to sort of show you what I was talking about with findings.  MANY of the findings are kind of BS, but there are a few legitimate ones there.   I was thinking about trying to address the legitimate ones in my fork.  Of course you'd be welcomed to merge them back here.

Thanks!

[apg-java DeveloperWorkbook.pdf](https://github.com/ldthomas/apg-java/files/6442301/apg-java.DeveloperWorkbook.pdf)

**Comment by ldthomas** [comment]
Yes, I would love to merge security fixes back into apg-java. If you could find a way to merge them into the master branch with a pull request, then you would show up as a collaborator, which would be fitting for as much work as you have done and as much help as you have given me. Not sure how you would do that, since your setup has such a different file structure but I'm confident we can find a way.

**Comment by ssill2** [comment]
Yeah I considered that too.  With the structure difference it would likely be a matter of hand porting the changes to my fork of master and then doing a pull request.

The other option, if you'd be willing, would be to simply allow your maven branch to be converted to maven by me doing a pull request.     So your maven branch could actually become a maven project and I could just do changes and do pull requests for you to pull back into your maven branch.   Given the fact your branch is named maven it would make the most sense for it to actually be a maven project :)  But it's up to you.  Given there aren't that many source files it wouldn't be hard to make the changes to my copy of master and generate a pull request there, but it would be kind of ugly.

My fu with git is no where near as strong as it is with svn, so maybe there is a non manual way to do it but it's not apparent to me right now.

**Comment by ldthomas** [comment]
Will the changes affect all the files or just a few? Will the examples still all work the same? 

Having a maven branch be a real maven project is appealing and I have thought of it. But there are a couple of problems. I'm not likely to be updating apg-java anytime soon, but if I did, keeping the maven branch in synch would be a problem. But the biggest problem is that it puts it under my control and I'm neither the author nor maven expert enough to maintain it. I think the creation and maintenance is best left as a fork owned by the author.

I'm also not the greatest expert with GitHub. Handling a pull request that only updates one branch and not master (or main) is beyond me. I could probably screw that one up big time.

I think a manual update is the only way to go. The maven branch was really only for review. I've always planned to merge it with master and delete it. There are a couple of possibilities. 1) you could save off your maven-ized fork and create new fork of the one and only master branch. Update the necessary files from your saved fork and create a pull request. Then put your stuff back in place. More work for you, but you get to be collaborator. 2) Just tell me which files have change, I clone your fork and manually replace the files and push the changes up to master. Easy for you, but you have to settle for a "thank you" in the docs.

**Comment by ssill2** [comment]
I'm not sure how many actual users of this project there are out there but I would be surprised if any of them would object to simply having your project being a maven project.   But it is your project and I respect you leaving it the way it is.

I'm guessing the changes would likely only be a few files.   I'll have to go do a more thorough scan to see what things I'd be inclined to fix from that report.  I know for a fact that some are very low hanging fruit and would be super easy and low risk changes.

I'm thinking the easiest thing to do will be:
1. have you merge your maven branch back into master.  You could create create a tag of the state of master before you do this if someone wanted to be able to pull the code as it was.
2. I'll delete my fork again and create a new fork from your master that has your maven branch merged in.
3. I'll create a branch in my forked copy that's converted to maven.  I'll work in this branch going forward.  If you changes in master I'll just have to watch for them and try to merge them.
4 If I make changes in my maven branch I'll make them in my copy of master and do a pull request for you to be able to pick them up.

**Comment by ldthomas** [comment]
Works for me. I'll get the tag, merge and delete done today. Then I'll be moving on to other things until I hear there is a merge request. I'm not familiar with Fortify but I looked at the report and kind of got the gist of it but couldn't identify any specific files and problems.

**Comment by ssill2** [comment]
Alright, I'll keep an eye out.  hopefully I only have to do the mavenize one more time lol  I'm assuming I can set up alerts to be notified if you make changes in master requiring me to go see about fulling your changes to my fork and then merging/porting changes to my maven branch.

The report is broken up in to sections, but the files and line numbers should be in there.  I have to do this for about 40 projects in order to submit our code to the government so I'm pretty familiar with Fortify at this point lol.  That scan is a worst case scenario.  All the options are turned on, such as finding code quality issues.  The other thing fortify asks when you create the scan profile is "is this a JEE application?" and "does this code run as root?"   I answer yes to both since I assume that is what the government does when we submit the code.  They also have fortify and I assume they run it against our code as well.  Thus why I scan the way I assume they are.   In my case I have a mixture of 40 projects, some of which are JEE apps, and some of which do run as root.   It's just easier to have fortify scan all 40 projects at one time vs. one for each project.

**Comment by ldthomas** [comment]
All done. The previous master is tagged V1.0 and the latest is tagged V1.1.0. Sorry to make you "maven-ize" this all over again. The good new is that I probably won't be making any changes anytime soon. As I said at the beginning, it's been a long time since I've done any Java development and I don't see any on the horizon any time soon. It's been great fun working with you.

**Comment by ssill2** [comment]
Likewise!   I'm glad to help revive this thing.  I've found it very useful in my current work and I think others probably could to.
Do consider converting the master to maven at some point.  I think others that find this because they search for "java abnf" like I did would probably appreciate it being a maven project.   But let's see how this goes for now.  Perhaps if you get enough votes for it being made a maven project it will sway you :)

I will likely be looking at possible changes this coming week and I'll ping you when I've got something.

Thanks again!
